### PR TITLE
Host user creation - Only update groups if needed

### PR DESCRIPTION
**Pull request by atburke**
This change updates host user creation to only update a host user's groups when they have changed (in the role). This allows nodes to skip acquiring a user lock most of the time, which fixes performance issues when a Teleport user `exec`s repeatedly.

Benchmarks from running `tsh ssh` 5 times concurrently are below.

Before:
| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `seq 5 \| parallel tsh ssh -l keeptemp node echo` | 9.338 ± 4.842 | 5.351 | 15.777 | 1.00 |

After:
| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `seq 5 \| parallel tsh ssh -l keeptemp node echo` | 367.2 ± 26.9 | 342.7 | 505.5 | 1.00 |

Changelog: Improved tsh ssh performance for concurrent execs

### CODE REVIEW

**Review by strideynet — APPROVED** [review]
Looks good - thanks for including some performance measurements. Definitely looks like it makes a big difference!

**Comment by atburke** [comment]
Friendly ping @fspmarshall

File: lib/srv/usermgmt.go [hunk 1]
```diff
@@ -276,6 +276,38 @@ func (u *HostUserManagement) UpsertUser(name string, ui *services.HostUsersInfo)
...
+				return nil, trace.Wrap(err)
+			}
+			currentGroups[group.Name] = struct{}{}
+		}
+
+		// Get the groups that the user should end up with, including the primary group.
+		finalGroups := make(map[string]struct{}, len(groups)+1)
+		for _, group := range groups {
+			finalGroups[group] = struct{}{}
+		}
```
[the remark this answers is not in the document]
Comment (atburke) [reply lib/srv/usermgmt.go]: `finalGroups` and `groups` aren't the same; `finalGroups` also includes the primary group. If we had just `groups`, we'd have to add and remove the primary group multiple times, which I think would be worse for readability than just having both containers.

**Comment by atburke** [comment]
Friendly ping @fspmarshall @rosstimothy

**Review by rosstimothy — APPROVED** [review]

**Pull request merged**